mm2src/coins/eth.rs

```diff
@@ -6179,3 +6190,755 @@ async fn get_eth_gas_details(
+                    let decoded = decode_contract_call(function, &tx_from_rpc.input.0)
+                        .map_to_mm(|err| ValidatePaymentError::TxDeserializationError(err.to_string()))?;
+
+                    let validation_params = ValidationParams {
+                        maker_address,
+                        etomic_swap_contract,
+                        token_id: args.token_id,
+                        amount: None,
+                    };
+                    validate_decoded_data(&decoded, &validation_params)?;
+
+                    let taker_address =
+                        addr_from_raw_pubkey(args.taker_pub).map_to_mm(ValidatePaymentError::InternalError)?;
+                    let htlc_params = ExpectedHtlcParams {
+                        swap_id,
+                        taker_address,
+                        token_address,
+                        taker_secret_hash: args.taker_secret_hash.to_vec(),
+                        maker_secret_hash: args.maker_secret_hash.to_vec(),
+                        time_lock: U256::from(args.time_lock),
+                    };
+                    decode_and_validate_htlc_params(decoded, 3, htlc_params)?;
+                },
+            },
+            EthCoinType::Eth | EthCoinType::Erc20 { .. } => {
+                return MmError::err(ValidatePaymentError::InternalError(
+                    "EthCoinType must be Nft".to_string(),
+                ))
+            },
+        }
+        Ok(())
+    }
+
+    async fn payment_status_v2(
+        &self,
+        swap_contract_addr: Address,
+        swap_id: Token,
+        contract: &Contract,
+        state_type: StateType,
+    ) -> Result<U256, PaymentStatusErr> {
+        let function_name = state_type.as_str();
+        let function = contract.function(function_name)?;
+        let data = function.encode_input(&[swap_id])?;
+        let bytes = self.call_request(swap_contract_addr, None, Some(data.into())).await?;
+        let decoded_tokens = function.decode_output(&bytes.0)?;
+        let state = decoded_tokens.get(2).ok_or_else(|| {
+            PaymentStatusErr::Internal("Payment status must contain 'state' as the 2nd token".to_string())
+        })?;
+        match state {
+            Token::Uint(state) => Ok(*state),
+            _ => Err(PaymentStatusErr::Internal(format!(
+                "Payment status must be Uint, got {:?}",
+                state
+            ))),
+        }
+    }
+
+    async fn spend_nft_maker_payment_v2_impl(
+        &self,
+        args: SpendNftMakerPaymentArgs<'_, Self>,
+    ) -> Result<SignedEthTx, TransactionErr> {
+        let contract_type = try_tx_s!(self.parse_contract_type(args.contract_type));
+        let etomic_swap_contract = try_tx_s!(self.parse_token_contract_address(args.swap_contract_address));
+        if args.maker_secret.len() != 32 {
+            return Err(TransactionErr::Plain(ERRL!("maker_secret must be 32 bytes")));
+        }
+        match self.coin_type {
```

Comment:
Again, this match statement can be simplified and extracted to a function, I don't want the swap functions to grow as it was with v1 implementations where it was very hard to read/review. Please look for other similar optimizations, e.g. `spend_nft_maker_payment_v2_impl`

Reply:
simplified spend function here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2084/commits/6edcdb9196b0839f7926fa5518926bf852d6bcf4